src/GitHub.App/Services/GistCreator.cs

```diff
@@ -0,0 +1,76 @@
+﻿using System;
+using System.ComponentModel.Composition;
+using System.Diagnostics;
+using System.Reactive.Linq;
+using System.Reactive.Threading.Tasks;
+using System.Threading.Tasks;
+using GitHub.Api;
+using GitHub.Extensions;
+using GitHub.Models;
+using GitHub.UI;
+using Octokit;
+using GitHub.Factories;
+using GitHub.Primitives;
+
+namespace GitHub.Services
+{
+    [Export(typeof(IGistCreator))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    public class GistCreator : IGistCreator
+    {
+        readonly ISelectedTextProvider selectedTextProvider;
+        readonly IConnectionManager connectionManager;
+        readonly IRepositoryHosts repositoryHosts;
+        readonly IUIProvider uiProvider;
+        readonly IApiClient apiClient;
+
+        [ImportingConstructor]
+        public GistCreator(ISelectedTextProvider selectedTextProvider, IConnectionManager connectionManager, 
+            IRepositoryHosts repositoryHosts, IUIProvider uiProvider, IApiClientFactory apiClientFactory)
+        {
+            this.selectedTextProvider = selectedTextProvider;
+            this.connectionManager = connectionManager;
+            this.repositoryHosts = repositoryHosts;
+            this.uiProvider = uiProvider;
+            this.apiClient = apiClientFactory.Create(HostAddress.GitHubDotComHostAddress);
+        }
+
+        public async Task<Gist> CreateFromSelectedText()
+        {
+            await EnsureLoggedIn();
+
+            var selectedText = selectedTextProvider.GetSelectedText().ToTask().Result;
+
+            // Show the Create Gist Window, and on Create button clicked execute the following
+            return await apiClient.CreateGist("NameWillBeEnteredInThePopup", true, selectedText)
+                .Catch<Gist, Octokit.NotFoundException>(_ => Observable.Return<Gist>(null))
+                .ToTask();
+        }
+
+        async Task EnsureLoggedIn()
+        {
+            var loggedIn = await connectionManager.IsLoggedIn(repositoryHosts);
+
+            if (!loggedIn)
+                loggedIn = await Login();
+
+            if (!loggedIn)
+                throw new InvalidOperationException("cannot create a gist if the user is not authenticated");
+        }
+
+        async Task<bool> Login()
+        {
+            var loggedInComplete = new TaskCompletionSource<bool>();
+
+            var authUI = uiProvider.SetupUI(UIControllerFlow.Authentication, null);
+            authUI.Subscribe(_ => { }, async () =>
+            {
+                var isLoggedIn = await connectionManager.IsLoggedIn(repositoryHosts);
+                loggedInComplete.SetResult(isLoggedIn);
+            });
+            uiProvider.RunUI();
```

Comment:
To clarify: the point is to split up the "am I logged in" logic from the actual gist creation logic. The login logic needs to happen at multiple points all across the extension (ever more as we add more functionality). Right now it's true that it's mostly triggered from the section and navigation items in the team explorer pages, which makes the current code a bit less generic than I would like.

I'm currently rewriting the UIController to allow for ui flows that trigger authentication dialogs automatically, so this flow will fit in that logic. In the meanwhile, the gist creation service should not care about whether it's logged in or not, just like all the other services don't care (see https://github.com/github/VisualStudio/blob/master/src/GitHub.App/Services/RepositoryCloneService.cs for an example). Assume we're logged in already when we get here.